### PR TITLE
Add non-overlay webview creation to notebook webview service and use for rendering html in positron notebooks

### PR DESCRIPTION
**Pull request by nstrayer**
This PR has three main parts to it as told by the (first) three commits:

1. Update the `IPositronNotebookWebviewService` to expose the `.createRawHtmlOutput()` method and allow that method to create standard `WebviewElement`s in addition to `OverlayWebview`s.  (Also adds the ability to create the webview without an active runtime due to runtime requirements preventing notebooks from showing html content until the kernel is connected.)
2. Add knowledge of the current language runtime attached to a notebook to the `PositronNotebookInstance` class.  (Also uses that knowledge to make the kernel badge more accurate)
3. Replace the existing rendering of HTML content in notebooks with the `IPositronNotebookWebviewService.createRawHtmlOutput()` and all the goodies that we get (and will get in the future) from that. 

### QA Notes

This should be a simple under-the-hood swap; nothing should look or work differently.

### CODE REVIEW

File: src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookInstance.ts [hunk 1]
```diff
@@ -510,7 +526,7 @@ export class PositronNotebookInstance extends Disposable implements IPositronNot
 	 * @returns True if the uri is the same as the notebook's uri, false otherwise.
 	 */
 	private _isThisNotebook(uri: URI): boolean {
-		return uri.toString() === this._input.resource.toString();
+		return uri.path === this._input.resource.path;
```
Comment (seeM) [inline src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookInstance.ts:529]: Could we use `isEqual` here ([src](https://github.com/posit-dev/positron/blob/3ed030d9c0cf5a7fb59824d290b8a596e99b3c60/src/vs/base/common/resources.ts#L360))?

File: src/vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewServiceImpl.ts [hunk 2]
```diff
@@ -302,7 +311,14 @@ window.onload = function() {
 	});
 };
 </script>`);
-		return new NotebookOutputWebview(id, runtime.runtimeMetadata.runtimeId, webview);
+
+		return new NotebookOutputWebview(
+			id,
+			runtime?.runtimeMetadata.runtimeId ?? '',
```
Comment (seeM) [inline src/vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewServiceImpl.ts:317]: Is this safe? At a quick glance, `INotebookOutputWebview.sessionId` is used in `WebviewPlotClient.metadata` which is used to store some state:

https://github.com/posit-dev/positron/blob/3ed030d9c0cf5a7fb59824d290b8a596e99b3c60/src/vs/workbench/contrib/positronPlots/browser/positronPlotsService.ts#L435

**Review by seeM — COMMENTED** [review]
Mostly looks good and same check from before looks correct:

image

Just one question about passing a blank `NotebookOutputWebview.sessionId`.

File: src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookInstance.ts (hunk 1 above)
Comment (nstrayer) [reply src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookInstance.ts:529]: We should. I didn't know that existed! I just checked for methods on the URI class! Thanks!

File: src/vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewServiceImpl.ts (hunk 2 above)
Comment (nstrayer) [reply src/vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewServiceImpl.ts]: Good catch. I totally forgot to go in and fix that. Updated so if you don't supply a runtime you have to supply some sort of session id string.

File: src/vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewServiceImpl.ts [hunk 3]
```diff
@@ -67,7 +67,12 @@ export class PositronNotebookOutputWebviewService implements IPositronNotebookOu
 		// HTML representation of the output.
 		for (const mimeType of Object.keys(output.data)) {
 			if (mimeType === 'text/html') {
-				return this.createRawHtmlOutput(output.id, runtime, output.data[mimeType]);
+				return this.createRawHtmlOutput({
+					id: output.id,
+					runtimeOrSessionId: runtime,
```
Comment (seeM) [inline src/vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewServiceImpl.ts:72]: Nit: Given that the runtime is currently only used for its `sessionId`, I wonder if we could make the arg just `sessionId`?

File: src/vs/workbench/contrib/positronNotebook/browser/notebookCells/NotebookHTMLOutput.tsx [hunk 4]
```diff
@@ -71,47 +72,56 @@ function acquireVsCodeApi(): { postMessage: (message: HTMLOutputWebviewMessage)
...
+		// Cleanup function that will be overwritten if the webview is created.
+		// If the effect gets cleaned up before the webview has been rendered it will
+		// set the disposed variable to true letting the webview creation know not to
+		// mount the webview.
+		let cleanup = () => { disposed = true; };
+
+		const buildWebview = async () => {
+			const webviewElement = await notebookWebviewService.createRawHtmlOutput({
+				id: 'notebook-html-output',
+				runtimeOrSessionId: notebookRuntime ?? 'positron.notebooks.output.render',
```
Comment (seeM) [inline src/vs/workbench/contrib/positronNotebook/browser/notebookCells/NotebookHTMLOutput.tsx:94]: I wonder if these should be unique? Might avoid issues down the line. In the console case both are UUIDs (I think).

**Review by seeM — DISMISSED** [review]
LGTM. One question in the comments + one issue: If I render a pandas dataframe in a positron notebook then press Cmd+R, I end up seeing two of them:

image

Could address that in a follow-up if it's quite involved.

File: src/vs/workbench/contrib/positronNotebook/browser/notebookCells/NotebookHTMLOutput.tsx (hunk 4 above)
Comment (nstrayer) [reply src/vs/workbench/contrib/positronNotebook/browser/notebookCells/NotebookHTMLOutput.tsx:94]: Oh I didnt realize that. I assumed that a runtime could have multiple webviews associated. I can make it unique!

File: src/vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewServiceImpl.ts (hunk 3 above)
Comment (nstrayer) [reply src/vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewServiceImpl.ts:72]: In the creation of the `WebviewInitInfo` there's a part that the extension ID is typed to be an `ExtensionIdentifier` and passing a string to it doesn't make typescript happy. https://github.com/posit-dev/positron/blob/836dc10789de912dd3c35a853813baeb3b3091d5/src/vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewServiceImpl.ts#L283-L284

**Comment by nstrayer** [comment]
I am pretty sure the double rendering thing isn't related as it seems to happen with any cell output type. I hadn't really known how to reproduce that before, so thanks for figuring it out!

File: src/vs/workbench/contrib/positronNotebook/browser/notebookCells/NotebookHTMLOutput.tsx (hunk 4 above)
Comment (nstrayer) [reply src/vs/workbench/contrib/positronNotebook/browser/notebookCells/NotebookHTMLOutput.tsx]: Now the id of the webview is the cell's `outputId` and the session id is a unique ID for the notebook instance when the runtime has not yet started up.

**Pull request merged**